doc: SUIT recovery button documentation (#19217)
nrfconnect/sdk-nrf: doc/nrf/app_dev/device_guides/nrf54h/ug_nrf54h20_suit_recovery.rst

    The value of ``SUIT_RECOVERY_APPLICATION_PATH`` can contain variables like ``${ZEPHYR_NRF_MODULE_DIR}``

#. When building the main application, set ``SB_CONFIG_SUIT_RECOVERY_APPLICATION_CUSTOM`` (or the Kconfig option name if a different one was chosen) to ``y``.

.. _ug_nrf54h20_suit_recovery_enter_recovery_request:

Entering recovery via are a request from the local domains
**********************************************************

> It is possible to enter recovery mode via a request sent from the local domains.
> This can be used for several purposes, such as:

Reviewer: ```suggestion
Recovery mode can be initiated through a request sent from local domains.  
This functionality is useful for several scenarios, including:
```

> * Recovering the device if the main firmware crashes and no JLink connection is available.
> * Using the recovery application as a "foreground update" application, allowing to save some space in the main application.
> * Debugging the recovery application.

Reviewer: ```suggestion
* Device recovery: Provides a recovery mechanism if the main firmware crashes and a J-Link connection is unavailable.  
* Foreground updates: Allows the recovery application to function as a "foreground update" application, reducing space requirements for the main application.  
* Application debugging: Facilitates debugging of the recovery application. 
```

> :ref:`Entering recovery via button press <ug_nrf54h20_suit_recovery_enter_via_button>` uses such requests.
> The code for this feature is located in the :file:`nrf/subsys/suit/app_tools/recovery_button` directory.
> It can be used as a reference design.
> 
> To request entering the recovery mode, the ``suit_foreground_dfu_required()`` function must be called.

Reviewer: ```suggestion
For additional details, see :ref:`Entering recovery via button press <ug_nrf54h20_suit_recovery_enter_via_button>`, which demonstrates similar request-based entry.

The implementation of this feature is available in the :file:`nrf/subsys/suit/app_tools/recovery_button` directory.  
This directory contains a reference design that can be adapted as needed.

To request the device to enter recovery mode, invoke the ``suit_foreground_dfu_required()`` function.  
```

Reviewer 2: Additional to these changes, line 252:

"For additional details, see :ref:`Entering recovery mode through button press <ug_nrf54h20_suit_recovery_enter_through_button>`, which demonstrates similar request-based entry."

Author: The suggested change has a bit different meaning than I intended - the recovery_button is a module which uses the described feature (entering recovery through a request - by the way, it uses it exactly, it does not use a "similar entry") - however it is not strictly a reference design/sample, it is a full-fledged feature. However, it is currently the only code which uses this feature, so I wanted to suggest to the documentation reader that if he wants to see an example, this module is the best place to start with.

I've modified the text a bit, please take a look now

> To exit recovery mode, the ``suit_boot_flags_reset()`` must be called from the recovery application.

Reviewer: ```suggestion
To exit recovery mode, invoke the ``suit_boot_flags_reset()`` function from the recovery application.  
```

> However - if the main application firmware is corrupted, the device will regardlessly enter recovery mode again on the next boot.

Reviewer: ```suggestion
If the main application firmware is corrupted, the device will re-enter recovery mode on the next boot, regardless of attempts to reset the boot flags.
```

"Enter recovery check" by running recovery firmware (companion image) before the main application
=================================================================================================

In some cases the device might need to check if the it should enter recovery mode before running the main application.
For example, the developer might want to enter recovery mode if a button is pressed during boot.

This can be done by running the recovery application as a companion image before the main application in the invoke path.
A different companion might be used for this purpose, but the recovery application is the most common choice.
It allows reusing the recovery firmware for multiple purposes, simplifying the system.

> The recovery (companion) application can check if the device should enter recovery mode and set the recovery flag using ``suit_foreground_dfu_required()`` if needed.
> It is important to only do this if the device is currently in the SUIT_BOOT_MODE_INVOKE boot mode - the current boot mode can be checked using ``suit_boot_mode_read``.
> If the device should remain in normal boot mode (the condidion is not met) or the device already is in recovery mode, the firmware should call ``suit_invoke_confirm(0)``.
> This will allow for further proccessing of the SUIT manifest by the Secure Domain Frimware and booting the main application.

Reviewer: ```suggestion
The recovery (companion) application can check if the device should enter recovery mode and set the recovery flag using ``suit_foreground_dfu_required()`` if needed.  
It is important to ensure this is only performed when the device is in the ``SUIT_BOOT_MODE_INVOKE`` boot mode.  
The current boot mode can be determined using the ``suit_boot_mode_read`` function.

If the device should remain in normal boot mode (for example, the condition is not met) or if the device is already in recovery mode, the firmware should call ``suit_invoke_confirm(0)``.  
This ensures that the SUIT manifest is processed further by the Secure Domain Firmware, allowing the main application to boot.
```

> To enable this feature, the following steps must be performed.
> This is assuming that the companion image is orchestrated by the APP_LOCAL_3 SUIT manifest - which will be true by default when using the recovery firmware as the companion.
> 
> 1. Add the appropriate checking code the recovery (companion) application in its startup code.
>    The following code snippet can be used:

Reviewer: ```suggestion
To enable this feature, the following steps must be performed.  
This assumes that the companion image is orchestrated by the ``APP_LOCAL_3`` SUIT manifest, which is the default setup when using the recovery firmware as the companion.

1. Add the appropriate checking code to the recovery (companion) application in its startup code.
   Use the following code snippet as a reference:
```

   .. code-block:: c

      static int should_enter_recovery_check(void)
      {
         suit_boot_mode_t mode = SUIT_BOOT_MODE_INVOKE_RECOVERY;
         suit_ssf_err_t err = SUIT_PLAT_SUCCESS;
         int ret = 0;

         err = suit_boot_mode_read(&mode);

         if (err != SUIT_PLAT_SUCCESS) {
            suit_invoke_confirm(-EPIPE);
            return -EPIPE;
         }

         if (mode == SUIT_BOOT_MODE_INVOKE) {
            if (/* add the condition here */) {
               err = suit_foreground_dfu_required();
            }
         }

         if (err != SUIT_PLAT_SUCCESS) {
            ret = -EPIPE;
         }

         (void)suit_invoke_confirm(ret);

         return ret;
      }

      SYS_INIT(should_enter_recovery_check, APPLICATION, CONFIG_APPLICATION_INIT_PRIORITY);

>    The code should be placed in the recovery (companion) application's source code.
>    To do this, create a module with this code and attach it to the Nordic provided recovery application or create a custom recovery application/companion image.

Reviewer: ```suggestion
   The code should be placed in the recovery (companion) application's source code.  
   To implement this, you can do one of the following:

   * Create a module with the code snippet and integrate it into the Nordic-provided recovery application.  
   * Create a custom recovery application or companion image and include the module there.
```

   However, otherwise, the manifest templates have to be modified as follows:

a. In the APP_LOCAL_3 manifest template, add the following code before the ``suit-directive-invoke`` directive on the companion image:

   .. code-block:: yaml

      - suit-directive-override-parameters:
          suit-parameter-invoke-args:
            suit-synchronous-invoke: True
            suit-timeout: 1000

b. In the root manifest template add the ``INSTLD_MFST`` component for the local manifest orchestrating the comapanion image.
   Make sure it is invoked before the main application images.
   If the modified template is based on the root manifest template from |NCS|, it will in most cases be enough to add it as the first manifst in the component list.
   The following code snippet is responsible for this in the default root manifest template:

   .. code-block:: yaml

      {{- component_list.append( app_recovery_local_component_index ) or ""}}
      - - INSTLD_MFST
        - RFC4122_UUID:
            namespace: {{ mpi_app_recovery_local_vendor_name }}
            name: {{ mpi_app_recovery_local_class_name }}

> .. _ug_nrf54h20_suit_recovery_enter_via_button:

Reviewer 2: ```suggestion
.. _ug_nrf54h20_suit_recovery_enter_through_button:
```

Entering recovery via button press
==================================

A common use case for entering recovery mode is to allow the device to enter recovery mode by pressing a button.
In NCS this use case is provided as a simple plug-and-play solution.
When it is enabled, the device will enter recovery mode if the button is pressed during boot.

.. note::
   If no modifications or additional modules are added to the recovery application the only way to exit the recovery mode is performing a device firmware upgrade.

There are two variants of this feature:

* Recovery button checked in the recovery application run as a companion image at boot stage (recommended).
* Recovery button checked in the main application.

The first variant is recommended as it allows to perform the check independently of the main application.

Recovery button checked in the recovery application
---------------------------------------------------

> In this variant,the recovery application is run as a companion image before the main application as part of the normal invoke path.

Reviewer 2: ```suggestion
In this variant, the recovery application is run as a companion image before the main application as part of the normal invoke path.
```

> It checks if the device already is in recovery mode.

Reviewer 2: ```suggestion
It checks if the device is already in recovery mode.
```

If it is, the recovery application proceeds with its normal operation allowing to recover firmware.

If the device is not in recovery mode, the recovery application checks if the specified button is pressed.
If the button is pressed, the recovery application sets the recovery flag and reboots the device.

> Otherwise it sends a confirmation message to the Secure Domain Firmware, which then halts the recovery application and proceeds with booting the main application.

Reviewer 2: ```suggestion
Otherwise, it sends a confirmation message to the Secure Domain Firmware, which then halts the recovery application and proceeds with booting the main application.
```

> To enable the feature add the following code to :file:`sysbuild/recovery.overlay` in the main application directory:

Reviewer 2: ```suggestion
To enable the feature, add the following code to the :file:`sysbuild/recovery.overlay` file in the main application directory:
```

   .. code-block:: dts

      / {
         chosen {
            ncs,recovery-button = &button0;
         };
      };

   Replace ``button0`` with the appropriate button node.

.. note::

>    In this option running the recovery application as a companion image is orchestrated by the root SUIT manifest.

Reviewer 2: ```suggestion
   In this option, running the recovery application as a companion image is orchestrated by the root SUIT manifest.
```

   Special care must be taken if the root manifest template is modified.
   If incorrect modifications are made to the root manifest, the button press feature might not work.
   This might lead to the device being unrecoverable without a JLink connection in case of a crash.

Recovery button checked in the main application
-----------------------------------------------

In this variant a check if the specified button is pressed is performed at an early stage when inside the main application firmware.
If the button is pressed, the recovery flag is set and the device reboots.

> To enable the feature in this variant add the following overlay to the main application's configuration:

Reviewer 2: ```suggestion
To enable the feature in this variant, add the following overlay to the main application's configuration:
```

    The value of ``SUIT_RECOVERY_APPLICATION_PATH`` can contain variables like ``${ZEPHYR_NRF_MODULE_DIR}``

#. When building the main application, set ``SB_CONFIG_SUIT_RECOVERY_APPLICATION_CUSTOM`` (or the Kconfig option name if a different one was chosen) to ``y``.

.. _ug_nrf54h20_suit_recovery_enter_recovery_request:

Entering recovery through a request from the local domains
**********************************************************

Recovery mode can be initiated through a request sent from local domains.
This functionality is useful for several scenarios, including:

* Device recovery: Provides a recovery mechanism if the main firmware crashes and a J-Link connection is unavailable.
* Foreground updates: Allows the recovery application to function as a "foreground update" application, reducing space requirements for the main application.
* Application debugging: Facilitates debugging of the recovery application.

> For an exemplary implementation, see :ref:`Entering recovery via button press <ug_nrf54h20_suit_recovery_enter_through_button>`, which implements such a request-based entry.

Reviewer 2: ```suggestion
For an exemplary implementation, see :ref:`Entering recovery through button press <ug_nrf54h20_suit_recovery_enter_through_button>`, which implements such a request-based entry.
```

The source code of this feature is available in the :file:`nrf/subsys/suit/app_tools/recovery_button` directory.

To request the device to enter recovery mode, invoke the ``suit_foreground_dfu_required()`` function.

To exit recovery mode, invoke the ``suit_boot_flags_reset()`` function from the recovery application.
If the main application firmware is corrupted, the device will re-enter recovery mode on the next boot, regardless of attempts to reset the boot flags.

"Enter recovery check" by running recovery firmware (companion image) before the main application
=================================================================================================

> In certain scenarios, the device might need to determine whether it should enter recovery mode before running the main application.

Reviewer: ```suggestion
In specific scenarios, the device must determine whether to enter recovery mode before starting the main application.
```

For example, a developer might require the device to enter recovery mode if a specific button is pressed during boot.

> This behavior can be achieved by running the recovery application as a companion image in the invoke path before the main application.
> While other companion images can be used for this purpose, the recovery application is the most common choice.

Reviewer: ```suggestion
To achieve this, you can configure the recovery application as a companion image in the invocation path preceding the main application.
Although other companion images can fulfill this purpose, the recovery application is the most commonly used option.
```

> Using the recovery application as a companion simplifies the system by allowing the recovery firmware to serve multiple purposes.

Reviewer: ```suggestion
Using the recovery application as a companion simplifies the system by enabling the recovery firmware to serve multiple purposes.
```

The recovery (companion) application can check if the device should enter recovery mode and set the recovery flag using ``suit_foreground_dfu_required()`` if needed.

> It is important to ensure this is only performed when the device is in the ``SUIT_BOOT_MODE_INVOKE`` boot mode.

Reviewer: ```suggestion
You must ensure that this is only performed when the device is in the ``SUIT_BOOT_MODE_INVOKE`` boot mode.
```

The current boot mode can be determined using the ``suit_boot_mode_read`` function.

If the device should remain in normal boot mode (for example, the condition is not met) or if the device is already in recovery mode, the firmware should call ``suit_invoke_confirm(0)``.
This ensures that the SUIT manifest is processed further by the Secure Domain Firmware, allowing the main application to boot.

> To enable this feature, the following steps must be performed.
> This assumes that the companion image is orchestrated by the ``APP_LOCAL_3`` SUIT manifest, which is the default setup when using the recovery firmware as the companion.

Reviewer: ```suggestion
To enable this feature, perform the following steps:

.. note::
   This assumes that the companion image is orchestrated by the ``APP_LOCAL_3`` SUIT manifest, which is the default 
   configuration when using the recovery firmware as the companion.

```

1. Add the appropriate checking code to the recovery (companion) application in its startup code.
   Use the following code snippet as a reference:

   .. code-block:: c

      static int should_enter_recovery_check(void)
      {
         suit_boot_mode_t mode = SUIT_BOOT_MODE_INVOKE_RECOVERY;
         suit_ssf_err_t err = SUIT_PLAT_SUCCESS;
         int ret = 0;

         err = suit_boot_mode_read(&mode);

         if (err != SUIT_PLAT_SUCCESS) {
            suit_invoke_confirm(-EPIPE);
            return -EPIPE;
         }

         if (mode == SUIT_BOOT_MODE_INVOKE) {
            if (/* add the condition here */) {
               err = suit_foreground_dfu_required();
            }
         }

         if (err != SUIT_PLAT_SUCCESS) {
            ret = -EPIPE;
         }

         (void)suit_invoke_confirm(ret);

         return ret;
      }

      SYS_INIT(should_enter_recovery_check, APPLICATION, CONFIG_APPLICATION_INIT_PRIORITY);

>    The code should be placed in the recovery (companion) application's source code.

Reviewer: ```suggestion
   You must place the code in the recovery (companion) application's source code.
```

   To implement this, you can do one of the following:

   * Create a module with the code snippet and integrate it into the Nordic-provided recovery application.
   * Create a custom recovery application or companion image and include the module there.

2. If using the default SUIT manifests and running the recovery image as the companion image, no modifications to the manifest templates are needed.
   Instead, set the :kconfig:option:`CONFIG_SUIT_INVOKE_APP_LOCAL_3_BEFORE_MAIN_APP` option in the recovery image.

>    This can be done by adding ``-Drecovery_CONFIG_SUIT_INVOKE_APP_LOCAL_3_BEFORE_MAIN_APP=y`` to the build command when building the recovery application.

Reviewer: ```suggestion
   To do so, add ``-Drecovery_CONFIG_SUIT_INVOKE_APP_LOCAL_3_BEFORE_MAIN_APP=y`` to the build command when building the recovery application.
```

>    However, otherwise, the manifest templates have to be modified as follows:

Reviewer: ```suggestion
   If you are not using the default SUIT maifests, you must modify the manifest templates as follows:
```

> a. In the APP_LOCAL_3 manifest template, add the following code before the ``suit-directive-invoke`` directive on the companion image:

Reviewer: ```suggestion
   a. In the ``APP_LOCAL_3`` manifest template, add the following code before the ``suit-directive-invoke`` directive on the companion image:
```

>    .. code-block:: yaml
> 
>       - suit-directive-override-parameters:
>           suit-parameter-invoke-args:
>             suit-synchronous-invoke: True
>             suit-timeout: 1000
> 
> b. In the root manifest template add the ``INSTLD_MFST`` component for the local manifest orchestrating the comapanion image.
>    Make sure it is invoked before the main application images.
>    If the modified template is based on the root manifest template from |NCS|, it will in most cases be enough to add it as the first manifst in the component list.
>    The following code snippet is responsible for this in the default root manifest template:

Reviewer: ```suggestion
     .. code-block:: yaml
  
        - suit-directive-override-parameters:
            suit-parameter-invoke-args:
              suit-synchronous-invoke: True
              suit-timeout: 1000

  b. In the root manifest template, add the ``INSTLD_MFST`` component for the local manifest orchestrating the comapanion image.
     You must ensure it is invoked before the main application images.
     If the modified template is based on the root manifest template from the |NCS|, it will in most cases be enough to add it as the first manifest in the component list.
     The following code snippet is responsible for this in the default root manifest template:
```

   .. code-block:: yaml

      {{- component_list.append( app_recovery_local_component_index ) or ""}}
      - - INSTLD_MFST
        - RFC4122_UUID:
            namespace: {{ mpi_app_recovery_local_vendor_name }}
            name: {{ mpi_app_recovery_local_class_name }}

.. _ug_nrf54h20_suit_recovery_enter_through_button:

Entering recovery mode through button press
===========================================

A common use case for entering recovery mode is to allow the device to enter recovery mode by pressing a button.

> In the |NCS| this use case is provided as a simple plug-and-play solution.

Reviewer: ```suggestion
In the |NCS|, this use case is provided as a simple plug-and-play solution.
```

When it is enabled, the device will enter recovery mode if the button is pressed during boot.

.. note::
   If no modifications or additional modules are added to the recovery application, the only way to exit the recovery mode is by performing a device firmware upgrade.

There are two variants of this feature:

> * Recovery button checked in the recovery application run as a companion image at boot stage (recommended).
> * Recovery button checked in the main application.

Reviewer: ```suggestion
* The recovery button checked in the recovery application running as a companion image during the boot stage (recommended).
* The recovery button checked in the main application.
```

> The first variant is recommended as it allows to perform the check independently of the main application.

Reviewer: ```suggestion
The first variant is recommended because it allows the check to be performed independently of the main application.
```

Recovery button checked in the recovery application
---------------------------------------------------

> In this variant, the recovery application is run as a companion image before the main application as part of the normal invoke path.
> It checks if the device is already in recovery mode.
> If so, the recovery application proceeds with its normal operation allowing to recover firmware.

Reviewer: ```suggestion
In this variant, the recovery application is executed as a companion image before the main application as part of the normal invocation path. 

The recovery application checks whether the device is already in recovery mode. 
If the device is in recovery mode, the recovery application continues its normal operation, enabling firmware recovery.
```

> If the device is not in recovery mode, the recovery application checks if the specified button is pressed.

Reviewer: ```suggestion
If the device is not in recovery mode, the recovery application checks whether the specified button is pressed. 
```

If the button is pressed, the recovery application sets the recovery flag and reboots the device.
Otherwise, it sends a confirmation message to the Secure Domain Firmware, which then halts the recovery application and proceeds with booting the main application.

To enable the feature, add the following code to the :file:`sysbuild/recovery.overlay` file in the main application directory:

   .. code-block:: dts

      / {
         chosen {
            ncs,recovery-button = &button0;
         };
      };

   Replace ``button0`` with the appropriate button node.

.. note::
   In this option, running the recovery application as a companion image is orchestrated by the root SUIT manifest.
   Special care must be taken if the root manifest template is modified.

>    If incorrect modifications are made to the root manifest, the button press feature may not work.
>    This may lead to the device being unrecoverable without a JLink connection in case of a crash.

Reviewer: ```suggestion
   Incorrect modifications to the root manifest can result in the button press feature not functioning correctly.
   This may cause the device to become unrecoverable without a JLink connection in the event of a crash.
```

Recovery button checked in the main application
-----------------------------------------------

> In this variant a check if the specified button is pressed is performed at an early stage when inside the main application firmware.
> If the button is pressed, the recovery flag is set and the device reboots.

Reviewer: ```suggestion
In this variant, a check is performed within the main application firmware at an early stage to determine if the specified button is pressed. 
If the button is pressed, the recovery flag is set, and the device reboots.
```

> To enable the feature in this variant, add the following overlay to the main application's configuration:

Reviewer: ```suggestion
To enable this feature in this variant, add the following overlay to the main application's configuration:
```

To achieve this, you can configure the recovery application as a companion image in the invocation path preceding the main application.
Although other companion images can fulfill this purpose, the recovery application is the most commonly used option.
Using the recovery application as a companion simplifies the system by enabling the recovery firmware to serve multiple purposes.

The recovery (companion) application can check if the device should enter recovery mode and set the recovery flag using ``suit_foreground_dfu_required()`` if needed.
You must ensure that this is only performed when the device is in the ``SUIT_BOOT_MODE_INVOKE`` boot mode.
The current boot mode can be determined using the ``suit_boot_mode_read`` function.

If the device should remain in normal boot mode (for example, the condition is not met) or if the device is already in recovery mode, the firmware should call ``suit_invoke_confirm(0)``.
This ensures that the SUIT manifest is processed further by the Secure Domain Firmware, allowing the main application to boot.

To enable this feature, perform the following steps:

.. note::

>    This assumes that the companion image is orchestrated by the ``APP_LOCAL_3`` SUIT manifest, which is the default 
>    configuration when using the recovery firmware as the companion.

Reviewer: ```suggestion
   This assumes that the companion image is orchestrated by the ``APP_LOCAL_3`` SUIT manifest, which is the default configuration when using the recovery firmware as the companion.
```

    The value of ``SUIT_RECOVERY_APPLICATION_PATH`` can contain variables like ``${ZEPHYR_NRF_MODULE_DIR}``

#. When building the main application, set ``SB_CONFIG_SUIT_RECOVERY_APPLICATION_CUSTOM`` (or the Kconfig option name if a different one was chosen) to ``y``.

Entering recovery via button press
**********************************

It is possible to enable a feature that allows the device to enter recovery mode by pressing a button.
This can be used for several purposes, such as:

* Recovering the device if the main firmware crashes and no JLink connection is available.
* Using the recovery application as a "foreground update" application, allowing to save some space in the main application.
* Debugging the recovery application.

When this feature is enabled, the device will enter recovery mode if the button is pressed during boot.

.. note::

>    To exit the recovery mode, a device firmware update must be performed.

Reviewer 3: Cold boot will not allow to leave this mode?

There are two variants of this feature:

* Recovery button checked in the recovery application run as a companion image at boot stage (recommended).
* Recovery button checked in the main application.

The first variant is recommended as it allows to perform the check independently of the main application.

Recovery button checked in the recovery application
===================================================

> In this variant the recovery application is run as a companion image before the main application when the device boots.

Reviewer 3: ```suggestion
In this variant, during the boot the recovery application is run as a companion image before the main application .
```

Author: I've modified it in a slightly different way, but I think it's better now

It checks if the device already is in recovery mode.

> If it is, the recovery application proceeds with its normal operation allowing to recover firmware.

Reviewer 3: ```suggestion
If so, the recovery application proceeds with its normal operation allowing to recover firmware.
```

Author: I think I'll let the tech writers decide on this one, it might slightly change the meaning

Reviewer 2: IMO, "If so" sounds more accurate.

"_If so, the recovery application proceeds with its normal operation, allowing the firmware to be recovered._"?

If the device is not in recovery mode, the recovery application checks if the specified button is pressed.
If the button is pressed, the recovery application sets the recovery flag and reboots the device.
Otherwise it sends a confirmation message to the Secure Domain Firmware, which then halts the recovery application and proceeds with booting the main application.

To enable the feature in this variant perform the following steps:

1. Add the following code to :file:`sysbuild/recovery.overlay` in the main application directory:

   .. code-block:: dts

      / {
         chosen {
            ncs,recovery-button = &button0;
         };
      };

   Replace ``button0`` with the appropriate button node.

2. Enable :kconfig:option:`CONFIG_SUIT_RECOVERY_BUTTON` in the *recovery application* configuration.

   When building from the command line, this can be done by adding ``-Drecovery_CONFIG_SUIT_RECOVERY_BUTTON=y`` to the build command.

.. note::
   In this option running the recovery application as a companion image is orchestrated by the root SUIT manifest.
   Special care must be taken if the root manifest template is modified.

>    If incorrect modifications are made to the root manifest, the button press feature might not work.

Reviewer 3: ```suggestion
   If incorrect modifications are made to the root manifest, the button press feature may not work.
```

>    This might lead to the device being unrecoverable without a JLink connection in case of a crash.

Reviewer 3: ```suggestion
   This may lead to the device being unrecoverable without a JLink connection in case of a crash.
```

Recovery button checked in the main application
===============================================

In this variant a check if the specified button is pressed is performed at an early stage when inside the main application firmware.
If the button is pressed, the recovery flag is set and the device reboots.

To enable the feature in this variant perform the following steps:

1. Add the following overlay to the main application's configuration:

   .. code-block:: dts

      / {
         chosen {
            ncs,recovery-button = &button0;
         };
      };

   Replace ``button0`` with the appropriate button node.

> 2. Enable :kconfig:option:`CONFIG_SUIT_RECOVERY_BUTTON` in the main application configuration.

Reviewer 4: Just thinking - since `SUIT_RECOVERY_BUTTON` already depends on DTS: `$(dt_chosen_enabled,$(DT_CHOSEN_NCS_RECOVERY_BUTTON))`, maybe we can state `default y` in that symbol too, so the DTS modification will be the only one required to enable this feature?

In other words - is there a practical case, in which the `ncs,recovery-button` is defined, yet the recovery button should not be checked?

Author: You are probably right, I've changed this in the other PR and removed the second step
